invokeai/app/services/model_install/model_install_base.py

```diff
@@ -0,0 +1,236 @@
+import traceback
+from abc import ABC, abstractmethod
+from enum import Enum
+from pathlib import Path
+from typing import Any, Dict, List, Optional, Union
+
+from pydantic import BaseModel, Field
+from pydantic.networks import AnyHttpUrl
+
+from invokeai.app.services.config import InvokeAIAppConfig
+from invokeai.app.services.events import EventServiceBase
+from invokeai.app.services.invoker import Invoker
+from invokeai.app.services.model_records import ModelRecordServiceBase
+from invokeai.backend.model_manager import AnyModelConfig
+
+
+class InstallStatus(str, Enum):
+    """State of an install job running in the background."""
+
+    WAITING = "waiting"  # waiting to be dequeued
+    RUNNING = "running"  # being processed
+    COMPLETED = "completed"  # finished running
+    ERROR = "error"  # terminated with an error message
+
+
+class UnknownInstallJobException(Exception):
+    """Raised when the status of an unknown job is requested."""
+
+
+ModelSource = Union[str, Path, AnyHttpUrl]
```

Comment:
This is done. The only hitch was that `wait_for_installs()` returns a Dict[ModelSource, ModelInstallJob]. and so I had to add a `__hash__` method to each of the model source types and play some tricks with the string representation so that when used in a string context the model source objects act like strings. I've not done this before, so please check that what I've done is kosher.

Reply:
New `ModelSource` looks good. The dunder `__hash__` workaround sounds pythonic to me 👍